### ISSUE TITLE
vec_copy potential issue

### ISSUE THREAD

**Issue by Monsieur-Verdoux** [juliuskoskela/c-vec#15]
Hi!

Thanks for sharing this project! I'm following along with the exercises now to learn about vectors and I noticed that in the function vec_copy line 57 you set 

`copy_size = src->alloc_size `

Should it not be 

`copy_size = dst->alloc_size`?

Also, line 53

`vec_new(dst, src->len, dst->elem_size);`

Should it have the same fail check as the other functions:

```
if (vec_new(dst, len, elem_size) < 0)
		return (-1);
```

Thanks!

**Comment by juliuskoskela** [comment]
> copy_size = src->alloc_size 
> 
> Should it not be
> 
> copy_size = dst->alloc_size?

Yeah I think you are correct. The if else statement should choose the alloc_size which is smaller. Or I think that has been the idea (it's a while since I wrote this code :D). It's arguable if it should just fail if dst is smaller than the memory we try to copy. This will quickly bring us to interesting API questions. In other languages we usually have "owning" types (like t_vec would be since it owns it's memory allocation) and reference types which refer to an owned type (often called slices or views). To take this API to the next level you would implement `t_vec_view` or similar which would enable slicing ranges from a vec to views and copying them instead.

> Should it have the same fail check as the other functions:
> 
> if (vec_new(dst, len, elem_size) < 0)
> 		return (-1);

Indeed!

Please make a PR :)

**Comment by Monsieur-Verdoux** [comment]
Thanks for the in-depth answer, will definitely look into the "owning" topic! 

PR made, my first ever :)

**Comment by Monsieur-Verdoux** [comment]
One more note: in the vec_append function lines 179 and 181 the `alloc_size` and `dst->alloc_size` are used respectively for the vec_resize, however that function expects `len` as a parameter, and in other functions, e.g. vec_push and vec_insert it was indeed `len` that was passed to vec_resize. This confusion may be due to the fact that in subject.md "Ex4" section and vec.h the resize function is prototyped as `static int vec_resize(t_vec *src, size_t target_size)`", while in the "Implementation" section and in vec.c it's shown as `int vec_resize(t_vec *src, size_t target_len)`. 

So what is the intended implementation? I imagine it would make more sense to use `len` over `alloc_size` for it, as is done in most of the functions?

Additionally, for vec_append and vec_prepend, can it be assumed that `dst->elem_size == src->elem_size`? It seems to me that it's necessary for the purpose of these functions. If so, then the allocation size could be assigned as `(dst->len + src->len) * src->elem_size;` and we could pass `dst->len + src->len` to vec_resize in line 179.
